lib/pry/input_completer.rb

```diff
@@ -239,4 +227,30 @@ def build_path(input)
     end
     return path, input
   end
+  
+    private
+
+    def self.fetch_obj_space_collection
+      (self.obj_spase_mutex ||= Mutex.new).synchronize do
```

Comment:
it's better to avoid locks if we can. thread-local state is preferable to a global mutex shared across >1 threads. on average though i don't think pry is ever run on more than one thread. i think is fine as it is but if Pry::InputCompleter was an instance, then the cache could live local to that instance as well, and every thread(which has a Pry instance as a local) would have its own cache.